n3fit/src/n3fit/io/writer.py

```diff
@@ -213,98 +218,156 @@
-        of the fit in order to fix some of the variables which would be, in principle,
-        be shared by several different history objects.
+        Initializes the writer for all replicas.
+
+        This is decoupled from the writing of the fit in order to fix some of the variables
+        which would be, in principle, be shared by several different history objects.
 
         Parameters
         ----------
-            `replica_number`
-                index of the replica
-            `pdf_object`
+            `replica_numbers`
+                indices of the replicas
+            `pdf_objects`
                 function to evaluate with a grid in x to generate a pdf
             `stopping_object`
                 a stopping.Stopping object
+            `all_chi2s`
+                list of all the chi2s, in the order: tr_chi2, vl_chi2, true_chi2
             `q2`
                 q^2 of the fit
             `timings`
                 dictionary of the timing of the different events that happened
         """
-        self.replica_number = replica_number
-        self.pdf_object = pdf_object
+        self.replica_numbers = replica_numbers
+        self.pdf_objects = pdf_objects
         self.stopping_object = stopping_object
         self.q2 = q2
         self.timings = timings
+        self.tr_chi2, self.vl_chi2, self.true_chi2 = all_chi2s
 
-    def write_data(self, replica_path_set, fitname, tr_chi2, vl_chi2, true_chi2):
+    def write_data(self, save_path, fitname, weights_name):
         """
-        Wrapper around the `storefit` function.
+        Save all the data of a fit, for all replicas.
 
         Parameters
         ----------
-            `replica_path_set`
-                full path for the replica, ex: `${PWD}/runcard_name/nnfit/replica_1`
+            `save_path`
+                path for the replica, ex: `${PWD}/runcard_name/nnfit`
             `fitname`
-                name of the fit
-            `tr_chi2`
-                training chi2
-            `vl_chi2`
-                validation chi2
-            `true_chi2`
-                chi2 of the replica to the central experimental data
+                name of the fit, ex: `Basic_runcard`
+            `weights_name`
+                name of the file to save weights to, if not empty
         """
-        # Check the directory exist, if it doesn't, generate it
-        os.makedirs(replica_path_set, exist_ok=True)
+        os.makedirs(save_path, exist_ok=True)
+
+        self.preprocessing = [
+            pdf_object.get_preprocessing_factors() for pdf_object in self.pdf_objects
+        ]
+        self.arc_lengths = [
+            vpinterface.compute_arclength(pdf_object).tolist() for pdf_object in self.pdf_objects
+        ]
+        self.integrability_numbers = [
+            vpinterface.integrability_numbers(pdf_object).tolist()
+            for pdf_object in self.pdf_objects
+        ]
```

Comment:
I think these should be better handled under **one single loop**; it's basically a difference between $\mathcal{O}(n)$ and $\mathcal{O}(3n)$ in time complexity.

Reply:
It shouldn't matter for performance, but it's maybe a bit cleaner, so I changed it.